events/EventQueue.h

```diff
@@ -167,8 +167,8 @@ class EventQueue : private mbed::NonCopyable<EventQueue> {
      *  @return         true  if event was successfully cancelled
      *                  false if event was not cancelled (invalid queue or executing already begun)
      */
-    template<typename... Args>
-    bool cancel(UserAllocatedEvent<Args...> *event)
+    template<typename F, typename A>
+    bool cancel(UserAllocatedEvent<F, A> *event)
```

Comment:
Original code wasn't _technically_ wrong, but I can see why ARMC5 might have got confused.

We're not actually instantiating anything here. `UserAllocatedEvent` takes two template arguments, and we're declaring something that could potentially give it variable.

There should only be a parameter mismatch if this was instantiated with `Args...` having anything other than 2 arguments. And that would never happen by function parameter deduction (there are no `UserAllocatedEvent`s with anything other than 2 arguments to pass). It should only barf if explicitly called as `cancel<int,int,int>(nullptr)`.

Presumably ARMC5 has some logic that crosschecks template arguments and parameters when a template is declared, and that's activating here, and thinking "variadic" doesn't match "2". That check may have been reasonable before variadic arguments. It catches this:

    class Foo<typename A, typename B>;

    template <typename A>
    bool cancel(Foo<A> *event);

I don't actually know whether that is legal; and if it isn't, is a compiler is required to diagnose it if cancel is never instantiated?

Anyway, this is an example of the sort of ARMC5 glitch we're going to keep hitting. Its C++11 implementation is incomplete, and buggy. We'll have to keep working around it.

Reply:
> Anyway, this is an example of the sort of ARMC5 glitch we're going to keep hitting. Its C++11 implementation is incomplete, and buggy. We'll have to keep working around it.

😭 👍

Reply:
I have section _"11.13 C++11 supported features"_ from the ARM C 5.06 manual printed to 3 sheets of A4 and pinned to the side of my cubicle.

Recommended reading for anyone thinking of pushing their C++11 luck.